### PR TITLE
Convert Dispatch API tutorial to Task.

### PR DESCRIPTION
**Pull request by tbedford**
## Description

Tutorials should be used for use cases. 

This PR experiments with using Tasks feature to replace a non-use case tutorial with a Task.

I also added a couple of tasks not used in this Task config, but will be useful for other Task configs.

@mheap @marklewin would be interested to know what you think on the Task approach!

@pardel @hughhopkins @martyndavies also interested in your thoughts. For this type of tutorial (non-use case) is this a better approach? Thanks!

## Feedback/Thoughts

1. Should prereqs be a task? Before, this content was a partial. The advantage being you could pull that partial into all tutorials that had the same prereqs without duplicating content. Writing the prereqs as a task solves the reuse issue, but should each prereq actually be a task?
2. Steps 4. Generate a JWT for Facebook validation and 5. Link your Facebook Page to your Nexmo account raise an interesting question. Although they are separate *steps*, they are *not* independent of each other in that Step 4 is only useful for Step 5. So should a Task be thought of as a step or a possible multi-step piece of reusable content? (**Update:** I've combined into one Task so we can see what we think).

**NOTE:** And see comments below for more feedback...

### CODE REVIEW

**Comment by martyndavies** [comment]
This is an interesting approach, and not one that I'm averse to either. From a reusable component point of view this would be handy when creating 'multiple' tasks for a product that require the same steps to be completed.

I think what stands out most here is the additional index of items within the 'task' that I think better helps the user to locate the specific area of interest/problem they're trying to solve when looking to achieve this specific task. It feels better than scrolling, but I think that there might be a need to look at how this is done from a design perspective so as to not feel too bunched up, or lead the user to miss elements due to right-hand column 'blindness'.

**Comment by mheap** [comment]
I've asked @acaciadealmeida for her thoughts on how it's presented

**Comment by tbedford** [comment]
I agree Mark. I don't like having more than one step in a "step"/Task. I also think it breaks the flow of a step-by-step. It sticks out in my mind. I do think of each task as a step, as I think that makes most sense from the user's perspective. From our perspective we are not quite as efficient, as we should be able to reuse a self-contained task, as opposed to two steps that are actually dependent Tasks. Perhaps the terminology needs to change. The Task is a list of Steps? I think the sub-task idea could work, but does complicate things. I would prefer to keep it simple and just have a Task that consists of a set of Steps. Each Step is as self-contained as possible. Sometimes (but probably not that often), we might get a situation where you have two steps where the first step is meaningless without the second step, and so you lose some re-use (which I can live with), but on the plus side the user sees a logical sequence of steps to carry out.

File: _tasks/olympus/prereqs.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+---
+title: Comply with prerequisites
```
[the remark this answers is not in the document]
Comment (tbedford) [reply _tasks/olympus/prereqs.md]: Yes, I was going to ask you for a better word for that! Thanks! :D

File: _tasks/olympus/prereqs.md [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+description: You should ensure you having the necessary prerequisites in place.
+---
+
+# Comply with prerequisites
+
+You should ensure you having the necessary prerequisites in place:
```
[the remark this answers is not in the document]
Comment (tbedford) [reply _tasks/olympus/prereqs.md]: I will fix this babble today. BTW do you have any thoughts on how we should deal with prerequisites? Should each prerequisite be a Task? Should we have a special prerequisite step? My thinking is that nearly all Tasks (a sequence of Steps) will have prerequisites, so we should build prerequisites into the structure of a Task (thinking of it as perhaps a special Step, or pre-Step). In DITA you have <prereq> as part of a Task. (https://docs.oasis-open.org/dita/v1.2/os/spec/common/task2.html)

**Comment by marklewin** [comment]
@tbedford Just to make sure we're on the same page: I like to think of the thing that the user is trying to accomplish (Send a SMS, play audio into a call, etc) as the "Task" and the individual actions that they must perform to accomplish that Task as a "Step". So one Task comprises several Steps. I think this is the DITA way of doing things too ... see [here](https://www.oxygenxml.com/dita/styleguide/webhelp-feedback/Artefact/Topics_and_Information_Types/c_Difference_between_Processes_Procedures_and_Tasks.html).

In terms of prerequisites, I think it would work best if we presented them in a checklist format as "Before You Begin" or similar, linking out to further details where appropriate. This would make them appear in the list as a Step (the first one), but I don't really consider it as such. To me, the Steps are specific to the actual Task so that anyone who wants to accomplish a specific task must complete ALL of the Steps. Some of the prereqs might already be met, so not every reader will have to action all of them. For that reason, I wouldn't recommend breaking the prereqs out into individual steps (and also because there might be half a dozen or more before you get to the "meat", as it were). Does that make sense?

Regarding Sub-Tasks (which should technically be called Sub-Steps according to our nomenclature but which sounds absolutely ridiculous 😬 ). I agree that the added complexity isn't worth the effort. But I think we can include several actions within a step, so long as they are logically related. For example, I'd hate to see a separate Step for "Click OK" but would be very happy to see something like "Click the Blah tab, select 'Bar' from the 'Foo' dropdown, enter your inside leg measurement and then click OK" all listed as bullets (or otherwise identified as separate actions) within the "Configure trousers" step.

I guess this is a very long and rambly way of saying I think we're in agreement, with the possible exception of listing each individual prereq as a separate step!

**Comment by tbedford** [comment]
It sounds like we are on the same page. I like the DITA approach of a Task consisting of Steps. I also definitely agree prereqs should not be split to multiple Tasks. I think prereqs should be in the `task.yml` file, probably between `introduction:` and `tasks:`. So, there should be an optional `prereqs:` field. We could delineate or identify the prereqs in some way, so that it is distinct from a normal step. Perhaps an icon of some type, with a footprint icon for a step? 

Good point about simple actions such as Click OK. I agree. For me it's about what I can reasonably expect the reader to digest in a single step. I prefer less rather than more generally, as with a longer step people can get lost as to where they are.

**Comment by marklewin** [comment]
Yes, I love the idea of making it obvious that the prereqs are separate with an icon or some other visual indicator. I really like the expanded descriptions of each step in the prev/next step area too. Looks great!

**Comment by mheap** [comment]
> I also definitely agree prereqs should not be split to multiple Tasks

I'm not sure on this. In this example people may already have a M&D application that they're using for WhatsApp but they want to add FB messenger support. The `link a FB page` prereq is important here, but could be missed if it's part of a large "prerequisites" section

> The user doesn't really need to have a description of the task before clicking on it. But does need the description of the step when the step is presented.

We've had feedback that the description under prev/next is useful - `I like how you have a back and forth mechanism with previews of the next and previous step`

> As the title is given along with description we should simply reuse that for the title of the topic, rather than having to yet again add the same text with a # in front of it.

Would this be too limiting if we wanted to embed tooling e.g. instead of sending people off to a static page to link a FB account we could embed it in the step as a component. I've been thinking about doing this too, so it may be worth trying and we can remove later if it is too limiting

> Yes, I love the idea of making it obvious that the prereqs are separate with an icon or some other visual indicator.

If we add a `prereqs:` section to the YAML config we could collapse them by default in the sidebar or add an icon to differentiate them

**Comment by tbedford** [comment]
@mheap: I think those latest changes broke something?

**Comment by mheap** [comment]
Yep they broke the build. I'll fix it up when I finish off the amendments

**Comment by mheap** [comment]
@tbedford @marklewin Prereqs section added - let me know what you think

**Comment by tbedford** [comment]
Perhaps even simpler by not having the prereqs on the right hand menu at all, but as a checklist on the first page under introduction and before the first step?

**Comment by mheap** [comment]
How would a consumer see the prereqs if they land on a page part way through a task via Google? We can't assume that they start at the beginning every time

**Comment by tbedford** [comment]
I think that's a slightly different issue. A "home" type button between Prev and Next would fix that.

**Comment by mheap** [comment]
I'd like them on every page I think. The home button requires people to click on them.

If we have composable steps that are shown as a single item in the list of tasks, how do people see which they need to do and which they don't? e.g. they may choose to skip create a Nexmo account, and go straight to application creation

**Comment by tbedford** [comment]
OK, if we do that can we make a clear prereqs is a different thing to the Steps. At the moment it looks like, at first glance, those are the steps for the prereqs. i.e. Prereqs -> Steps -> numbered steps. So if possible can we have, in the right-hand menu, Prereqs framed in a box, and Steps framed in a box, so it's completely clear these are separate things?

**Comment by tbedford** [comment]
Note that I think there is now too much info on the prereqs page - I think a simple list of prereqs is fine. If the user needs to know more they can click on the prereq to expand or go to that task.

Also, how do I link to a task? `[webhook is set](/_tasks/olympus/configure-webhooks)` doesn't seem to work.

**Comment by mheap** [comment]
@tbedford I've just pushed some style changes to the prerequisites page. Does that look ok to you?

You can't currently link to a single step - that'll be some followup work once this is merged

**Comment by tbedford** [comment]
Yes, that looks great!

**Comment by mheap** [comment]
This branch has had some `git` surgery applied. To make your local version work again run the following commands:

```
git checkout tony-dispatch-task
git fetch
git reset --hard origin/tony-dispatch-task
```

**Comment by mheap** [comment]
@tbedford I think we're finally there! Showing up in the sidebar, with the correct URL

https://nexmo-developer-pr-1672.herokuapp.com/dispatch/tutorials/sending-facebook-message-with-failover/introduction

**Comment by tbedford** [comment]
LGTM.

**Comment by mheap** [comment]
Just realised that we're missing canonical URL support which needs adding before this is merged. I'll add that today

**Pull request merged**